### PR TITLE
APP-3206: Revamping all BDK services

### PR DESCRIPTION
**Pull request by symphony-hong**
### Ticket
[APP-3206](https://perzoinc.atlassian.net/browse/APP-3206)

### Description
Revamping all BDK services
Breaking changes:
- `MessageService#getMessages(V4Stream, Instant, Integer, Integer)` -> `MessageService#listMessages(V4Stream, Instant, PaginationAttribute)`
- `MessageService#getMessages(String, Instant, Integer, Integer)` -> `MessageService#listMessages(String, Instant, PaginationAttribute)`
- `MessageService#getMessagesStream(V4Stream, Instant, Integer, Integer)` -> `MessageService#listAllMessages(V4Stream, Instant, StreamPaginationAttribute)`
- `MessageService#getMessagesStream(String, Instant, Integer, Integer)` -> `MessageService#listAllMessages(String, Instant, StreamPaginationAttribute)`
- `MessageService#blastMessage(List<String>, Message)` -> `MessageService#send(List<String>, Message)`
- `MessageService#getMessageIdsByTimestamp(String, Instant, Instant, Integer, Integer)` -> `MessageService#listMessageIdsByTimestamp(String, Instant, Instant, PaginationAttribute)`
- `MessageService#getMessageIdsByTimestampStream(String, Instant, Instant, Integer, Integer)` -> `MessageService#listAllMessageIdsByTimestamp(String, Instant, Instant, StreamPaginationAttribute)`
- `PresenceService#getAllUsersPresence(Long, int)` -> `PresenceService#listPresencesOfAllUsers(Long, Integer)`
- `SignalService#listSignals(Integer, Integer)` -> `SignalService#listSignals(PaginationAttribute)`
- `SignalService#listSignalsStream(Integer, Integer)` -> `SignalService#listAllSignals(StreamPaginationAttribute)`
- `SignalService#listSignalsStream()` -> `SignalService#listAllSignals()`
- `SignalService#listSubscribers(String, Integer, Integer)` -> `SignalService#listSubscribers(String, PaginationAttribute)`
- `SignalService#listSubscribersStream(String, Integer, Integer)` -> `SignalService#listAllSubscribers(String, StreamPaginationAttribute)`
- `SignalService#listSubscribersStream(String)` -> `SignalService#listAllSubscribers(String)`
- `StreamService#getStreamInfo(String)` -> `StreamService#getStream(String)`
- `UserService#searchUserByIds(List, Boolean, Boolean)` -> `UserService#listUsersByIds(List, Boolean, Boolean)`
- `UserService#searchUserByIds(List)` -> `UserService#listUsersByIds(List)`
- `UserService#searchUserByEmails(List, Boolean, Boolean)` -> `UserService#listUsersByEmails(List, Boolean, Boolean)`
- `UserService#searchUserByEmails(List)` -> `UserService#listUserByEmails(List)`
- `UserService#searchUserByUsernames(List, Boolean)` -> `UserService#listUsersByUsernames(List, Boolean)`
- `UserService#searchUserByUsernames(List)` -> `UserService#listUsersByUsernames(List)`
- `UserService#searchUserBySearchQuery(UserSearchQuery, Boolean)` -> `UserService#listUsersBySearchQuery(UserSearchQuery, Boolean)`
- `UserService#searchUserBySearchQuery(UserSearchQuery, Boolean, PaginationAttribute)` -> `UserService#listUsersBySearchQuery(UserSearchQuery, Boolean, PaginationAttribute)`
- `UserService#searchAllUsersBySearchQuery(UserSearchQuery, Boolean)` -> `UserService#listAllUsersBySearchQuery(UserSearchQuery, Boolean)`
- `UserService#searchAllUsersBySearchQuery(UserSearchQuery, Boolean, StreamPaginationAttribute)` -> `UserService#listAllUsersBySearchQuery(UserSearchQuery, Boolean, StreamPaginationAttribute)`

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated

### CODE REVIEW

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java [hunk 1]
```diff
@@ -213,17 +274,17 @@ public V4Message send(@Nonnull String streamId, @Nonnull Message message) {
    * @param streamIds the list of stream IDs to send the message to
    * @param message   the message to be sent
    * @return a {@link V4MessageBlastResponse} object containing the details of the sent messages
-   * @see <a href="https://developers.symphony.com/restapi/reference#create-message-v4">Create Message v4</a>
+   * @see <a href="https://developers.symphony.com/restapi/v20.9/reference#blast-message">Blast Message</a>
    */
-  public V4MessageBlastResponse blastMessage(@Nonnull List<String> streamIds, @Nonnull Message message) {
+  public V4MessageBlastResponse sendBlast(@Nonnull List<String> streamIds, @Nonnull Message message) {
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java:279]: Why not just `send(List<String>, Message)`?

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java [hunk 2]
```diff
@@ -364,55 +426,85 @@ public V4Message getMessage(@Nonnull String messageId) {
...
+   * Fetches message ids using timestamp and returns in {@link Stream} with default chunk size and total size equals 100.
+   *
+   * @param streamId the ID of the stream where to fetch messages.
+   * @param since    optional instant of the first required messageId.
+   * @param to       optional instant of the last required messageId.
+   * @return a {@link Stream} containing the messageIds.
+   * @see <a href="https://developers.symphony.com/restapi/reference#get-message-ids-by-timestamp">Get Message IDs by Timestamp</a>
+   */
+  @API(status = API.Status.EXPERIMENTAL)
+  public Stream<String> listAllMessageIdsByTimestampStream(@Nonnull String streamId, @Nullable Instant since,
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java:500]: Stil `Stream` at the end?

**Review by symphony-thibault — COMMENTED** [review]
Could we track all breaking changes (method `x` renamed to `y`) in the PR description? So that it will be available in the next release note.

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java [hunk 3]
```diff
@@ -364,55 +426,85 @@ public V4Message getMessage(@Nonnull String messageId) {
...
+   * @param since      optional instant of the first required messageId.
+   * @param to         optional instant of the last required messageId.
+   * @param pagination The chunkSize and totalSize for pagination.
    * @return a {@link Stream} containing the messageIds.
    * @see <a href="https://developers.symphony.com/restapi/reference#get-message-ids-by-timestamp">Get Message IDs by Timestamp</a>
    */
-  public Stream<String> getMessageIdsByTimestampStream(@Nonnull String streamId, Instant since, Instant to,
-      Integer chunkSize, Integer totalSize) {
+  @API(status = API.Status.EXPERIMENTAL)
+  public Stream<String> listAllMessageIdsByTimestampStream(@Nonnull String streamId, @Nullable Instant since,
```
Comment (symphony-elias) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java:482]: Please remove trailing `Stream` in `listAllMessageIdsByTimestampStream`.

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java (hunk 2 above)
Comment (symphony-hong) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java:500]: I'm not sure about that. That's why I'm still annotated it as `EXPERIMENTAL`

Comment (symphony-thibault) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java]: _listAllMessageIdsByTimestamp_**Stream**() :)

Comment (symphony-hong) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/MessageService.java]: Oops 😅

**Review by symphony-thibault — APPROVED** [review]
LGTM 👍

**Review by symphony-youri — COMMENTED** [review]
I'm a bit late but I'll add some comments here to keep track of them

in user service we have many methods to search for a user, maybe having just one with a search parameters object would be simpler (there is already one like that takes a UserSearchQuery)

listUsersDetail -> is probably not going to return all users right? just the first 100? (cf the javadoc comment there)

methods often have 'OfUser' suffix, isn't it redundant? for a user service and a method 
getAvatarFromUser that takes a userid as a param you kind of expect it to be for the user (also here it is not consistent fromUser/OfUser)

SignalService vs UserService:
subscribeSignal / addRoleToUser
subscribeSignal -> subscribeUserToSignal would be more consistent (this is an open question) or just simply addRole and we can infer things based on the parameters

ApplicationService -> sometimes we use App sometimes Application

DatafeedService, we call that a service but it is a bit different from the other services we have

**Pull request merged**